app/src/test/java/org/mozilla/fenix/wallpapers/WallpaperDownloaderTest.kt

```diff
@@ -0,0 +1,134 @@
+    private val portraitResponseBodySuccess = Response.Body(wallpaperBytes.byteInputStream())
+    private val landscapeResponseBodySuccess = Response.Body(wallpaperBytes.byteInputStream())
+    private val mockPortraitResponse = mockk<Response>()
+    private val mockLandscapeResponse = mockk<Response>()
+    private val mockClient = mockk<Client>()
+
+    private val dispatcher = UnconfinedTestDispatcher()
+
+    private val wallpaperCollection = Wallpaper.Collection(
+        name = "collection",
+        heading = null,
+        description = null,
+        learnMoreUrl = null,
+        availableLocales = null,
+        startDate = null,
+        endDate = null
+    )
+
+    private lateinit var downloader: WallpaperDownloader
+
+    @Before
+    fun setup() {
+        downloader = WallpaperDownloader(tempFolder.root, mockClient, dispatcher)
+    }
+
+    @Test
+    fun `GIVEN that request is successful WHEN downloading THEN file is created in expected location`() = runTest {
+        val wallpaper = generateWallpaper()
+        val portraitRequest = Request(
+            url = "$remoteHost/${wallpaper.collection.name}/${wallpaper.name}/portrait.png",
+            method = Request.Method.GET
+        )
+        val landscapeRequest = Request(
+            url = "$remoteHost/${wallpaper.collection.name}/${wallpaper.name}/landscape.png",
+            method = Request.Method.GET
+        )
+        every { mockPortraitResponse.status } returns 200
+        every { mockLandscapeResponse.status } returns 200
+        every { mockPortraitResponse.body } returns portraitResponseBodySuccess
+        every { mockLandscapeResponse.body } returns landscapeResponseBodySuccess
+        every { mockClient.fetch(portraitRequest) } returns mockPortraitResponse
+        every { mockClient.fetch(landscapeRequest) } returns mockLandscapeResponse
+
+        downloader.downloadWallpaper(wallpaper)
+
+        val expectedPortraitFile = File(tempFolder.root, "wallpapers/${wallpaper.name}/portrait.png")
+        val expectedLandscapeFile = File(tempFolder.root, "wallpapers/${wallpaper.name}/landscape.png")
+        assertTrue(expectedPortraitFile.exists() && expectedPortraitFile.readText() == wallpaperBytes)
+        assertTrue(expectedLandscapeFile.exists() && expectedLandscapeFile.readText() == wallpaperBytes)
+    }
+
+    @Test
+    fun `GIVEN that request fails WHEN downloading THEN file is not created`() = runTest {
+        val wallpaper = generateWallpaper()
+        val portraitRequest = Request(
+            url = "$remoteHost/${wallpaper.collection.name}/${wallpaper.name}/portrait.png",
+            method = Request.Method.GET
+        )
+        val landscapeRequest = Request(
+            url = "$remoteHost/${wallpaper.collection.name}/${wallpaper.name}/landscape.png",
+            method = Request.Method.GET
+        )
+        every { mockPortraitResponse.status } returns 400
+        every { mockLandscapeResponse.status } returns 400
```

Comment:
appreciate the error handling validation. is it only 400s we have to worry about?

Reply:
We could potentially add more robust error-handling down the line but tbh I don't really see the point without telemetry or UX to support it. So AFAIK yes. Our `Response` only buckets statuses into successes and failures as seen [here](https://searchfox.org/mozilla-mobile/rev/e935d8b83b94a8d9996f1fa980409457bdcadb9c/android-components/components/concept/fetch/src/main/java/mozilla/components/concept/fetch/Response.kt#142)

Reply:
The underlying code is just checking `response.isSuccess` for now